Opera ticket zoe contract

DavidBruant opened the pull request:
Closes #543

To run the test independently, run:
`tape -r esm 'test/unitTests/contracts/test-operaConcertTicket.js' | tap-spec`
(in zoe package and with `./node_modules/.bin` in your `$PATH`)

I'm currently, stuck with a weird error
```
payment not found: zoeInvite
    FAILED ASSERTION false
(node:14889) UnhandledPromiseRejectionWarning: Error: payment not found: (a string)
See console for error data.
    at Object.complain (/home/david/projects/agoric-sdk/packages/assert/src/assert.js:112:14)
    at fail (/home/david/projects/agoric-sdk/packages/assert/src/assert.js:131:20)
    at assert (/home/david/projects/agoric-sdk/packages/assert/src/assert.js:155:5)
    at /home/david/projects/agoric-sdk/packages/ERTP/src/issuer.js:169:9
    at async Test.<anonymous> (/home/david/projects/agoric-sdk/packages/zoe/test/unitTests/contracts/test-operaConcertTicket.js:50:7)
```

but the payment is the return value of `await zoe.makeInstance`, so i'm fairly confident and surprised this wouldn't be caught by other tests.
I'm doing something else and will be back to it in a couple hours

katelynsills commented on packages/zoe/src/contracts/operaConcertTicket.js:
```diff
@@ -0,0 +1,53 @@
+      throw `TODO`
+    }
+  })
+  const auditoriumInvite = zoe.makeInvite(auditoriumSeat);
```
```suggestion
  const { invite: auditoriumInvite } = zoe.makeInvite(auditoriumSeat);
```
I think you're returning a record of invite, inviteHandle to Zoe when you should just return the invite. I think that's why it is saying it's not a payment - because it's a record, not a payment

DavidBruant replied:
oh what a trivial mistake :-/
i am grateful for the time you spent on this
i'll try to be more careful

katelynsills replied:
Happens to me all the time, and this exact thing might happen to a user, so it's all good. Thanks for working on this!

DavidBruant commented:
I've moved forward

My experience so far is that writing a contract is hard because of a combinaison of:
- ERTP+Zoe APIs very inconsistently return the object or a promise for the object
- what returns a promise and what does not is not really documented yet (and that may not be important in a world with the `~.` operator)
- error messages do not help with the difference
- in general, error messages lack context and as a consequence only provide a vague idea of what is going wrong
- tape reporting is noisy and so, the relevant debugging information is hard to get
- ERTP is hard, but i'm over the hill now, so it's easier, but Zoe adds another layer of concepts

DavidBruant commented on packages/zoe/src/contracts/operaConcertTicket.js:
```diff
@@ -0,0 +1,113 @@
+      // create an Zoe invite internally...
+      const {invite, inviteHandle} = zoe.makeInvite()
+      // ...and redeem it right away
+      return zoeService.redeem(
```
The approach i've taken is that since the mint is internal to the smart contract, the smart contract needs to create its own invitations to send the offers to Zoe. Thesmart  contract immediatly redeems the invites (and ignores the seat since it's irrelevant from the smart contract perspective)

@katelynsills I'm interested in your opinion on this approach
I didn't see an easier way. I wonder whether there could be a zoe contract facet method to make this easier (but is it really a relevant use case given the mint being internal is a temporary decision?)

DavidBruant replied:
Maybe that's where `makeEmptyOffer` is relevant

DavidBruant commented on packages/zoe/src/contracts/operaConcertTicket.js:
```diff
@@ -0,0 +1,113 @@
+        const ticketInviteHandle = ticketPaymentInviteHandles.get(ticketNumber)
+
+        // total bluff
+        return swap(ticketInviteHandle, buyerInviteHandle)
```
This line does not work (`Error: prior offer is unavailable`)
I'm currently stuck here, but i need to spend more time understanding how Zoe works

DavidBruant commented:
A first version of the contract is ready, so i'm marking this PR as ready for review

There are cosmetic aspects (lint/prettier/commit messages) that will need to be addressed, but the PR is ready for functional review

katelynsills commented on packages/zoe/src/contracts/operaConcertTicket.js:
```diff
@@ -0,0 +1,120 @@
+
+  // Create the internal ticket mint
+  const { issuer, mint, amountMath } = produceIssuer('Opera tickets', 'set');
+  await zoe.addNewIssuer(issuer, 'Auditorium');
```
I think Mark would prefer that we not use `async/await` in our examples.

DavidBruant replied:
i didn't know that. I'll fix it
Would it be worth adding an eslint warning (which is different than an eslint error) for this purpose?

erights replied:
> Would it be worth adding an eslint warning (which is different than an eslint error) for this purpose? [avoiding `await`]

Yes. We should start with a warning.

katelynsills commented on packages/zoe/src/contracts/operaConcertTicket.js:
```diff
@@ -0,0 +1,120 @@
+  const { issuer, mint, amountMath } = produceIssuer('Opera tickets', 'set');
+  await zoe.addNewIssuer(issuer, 'Auditorium');
+
+  // create Zoe helpers after zoe.addNewIssuer because of https://github.com/Agoric/agoric-sdk/issues/802
```
Good catch!

katelynsills commented on packages/zoe/src/contracts/operaConcertTicket.js:
```diff
@@ -0,0 +1,120 @@
+
+        salesPayouts.add(payout)
+
+        try{
```
Is this `try/catch` doing anything?

katelynsills commented on packages/zoe/src/contracts/operaConcertTicket.js:
```diff
@@ -0,0 +1,120 @@
+      },
+    });
+    const { invite, inviteHandle } = zoe.makeInvite(seat);
+    return { invite, inviteHandle };
```
I think we should only return the invite here. It should enable more promise pipelining.

katelynsills commented on packages/zoe/src/contracts/operaConcertTicket.js:
```diff
@@ -0,0 +1,120 @@
+  const salesPayouts = new Set()
+
+  const auditoriumSeat = harden({
+    async getSalesPayment(){
```
Instead of having this method for the entity selling the tickets to receive their payout, we should use the Zoe payout mechanism instead. The contract should retain the offerHandle of the original creator, then reallocate all the moola to that offerHandle as it is received. The contract will still have to create new invites to escrow tickets with Zoe so that the contract can reallocate them to the user, but at least we 1) guarantee offer safety for the seller of the tickets, and 2) don't have a duplicate mechanism for payouts. 

I think you are right that this "contract escrows things with Zoe to eventually give them to a user" pattern is going to be reused a lot, and we might want to make it a lot simpler.

DavidBruant replied:
One idea i had initially was that the Opera might want to get the moolas as they arrive (and not only once at the end like how the payout mechanism requires)
But i realize now that it unnecessarily complicates this first contract, so i'll ignore this requirement for now

Another problem i see with the payout is that the Opera would need to know it advance what it `want`s, but it cannot know in advance how many tickets will be sold and so, it cannot know whether it `want` 0 or 22 or 44 or 66 moolas and invite redeeming will only accept one proposal
If the Opera `want`s 22 moolas and 3 tickets get sold, the 44 leftover moolas can only remain in the contract
If the Opera `want`s 66 moolas and 1 ticket is sold, there is never enough moolas to perform a reallocation of the moolas that satisfied `want:` 66 moolas

i do not know how to solve this problem with Zoe. How to describe a proposal with an unknown extent?
(i'm starting to think that maybe 0 could work with a manual call to `.reallocate` with the final amount, i'll dig into that)

DavidBruant replied:
An alternative to all this would be:
- the contract mints the tickets + invites and passes all this to the Opera
- the Opera redeems invites with `give ticket want 22 moolas`
- the contract has kept the offerHandles and is able to track which ticket is still available with them
- The Opera uses all the individual payouts

With this alternative, the contract never redeems invites internally and never plays with payouts

DavidBruant replied:
To expand a bit on this idea. I was initially puzzled by the argument "guarantee offer safety for the seller of the tickets" because i felt that if the contract code was correct, offer safety was guaranteed. But this "if" is literally one of the things being defended against, so that's not an assumption that can be made

I feel like a contract that plays with invites and payouts internally cannot really guarantee offer safety since a bug may lead to invites not being properly redeemed with incorrect proposals/payments or payouts not being recovered and that would lead to assets being stuck in the contract which was one of the things offer safety was supposedly defending against

If i'm stretching this idea, I would go as far as considering that inside a zoe contract it shouldn't be possible to call `zoe.redeem`

erights replied:
> Another problem i see with the payout is that the Opera would need to know it advance what it wants, but it cannot know in advance how many tickets will be sold 

Yes, this is a lack of expressiveness that we absolutely need to fix. We started by thinking of this as a "divisible" offer, but then we got into roundoff hell. Instead, we're thinking of this a "multiples": Create the smallest atomic offer: "1 ticket for 1 ticket price" and place it inside a repetition factor of N. It acts like a set of N separate but identical offers, any number of which can be accepted.

Another missing expressiveness we expect to add is exclusive choice: "offer A or offer B but not both".

With one additional element, an "offset", these three together lets us express a piecewise linear approximation to an indifference curve, which is an interesting test of "adequate" generality.

Note of this will happen by Alpha.

DavidBruant commented on packages/zoe/test/unitTests/contracts/test-operaConcertTicket.js:
```diff
@@ -0,0 +1,222 @@
+    const { source, moduleFormat } = await bundleSource(operaConcertTicketRoot);
+
+    const installationHandle = zoe.install(source, moduleFormat);
+    const { invite: auditoriumInvite } = await zoe.makeInstance(installationHandle, harden({Buyer: moolaIssuer}), {
```
Yeah, that was my first use of keywords and it wasn't clear to me 
I'm changing the keywords to Money and Tickets
And i created a documentation issue about Keywords: https://github.com/Agoric/documentation/issues/102

katelynsills replied:
Ok, great! Thank you for making that issue. I'd be very interested in hearing your thoughts on how we can better explain keywords.

DavidBruant commented on packages/zoe/test/unitTests/contracts/test-operaConcertTicket.js:
```diff
@@ -0,0 +1,222 @@
+  
+    const {seat: {performExchange}, payout: payoutP} = await zoe.redeem(bobInvite, bobProposal, {Buyer: bobFirstPaymentForTicket})
+
+    await performExchange()
```
I tried and failed to use `t.throw` for a reason that remains a mystery to me, hence how the code is now (`await performExchange().then(() => t.fail('should throw')).catch(error => {})`)

I'm not satisfied with the solution, but i don't know how to do better

DavidBruant replied:
i managed to use `t.throw`

DavidBruant commented:
> I will do another pass after the linting errors are fixed - it was hard to view in my local VSCode environment with the errors, so I may have missed things.

I didn't have eslint enabled on VSCode so i didn't realize how annoying it might have been to read the code. I've enabled it and fixed the errors

DavidBruant commented on packages/zoe/test/unitTests/contracts/test-operaConcertTicket.js:
```diff
@@ -0,0 +1,222 @@
+The Opera is told about the show being sold out. It gets all the moolas from the sale`, async t => {  
+
+  // Setup initial conditions
+  const { mints, issuers: defaultIssuers, moola } = setup();
```
I've removed the call to `setup` altogether since a call to produceIssuer + destructuring is readable enough

katelynsills replied:
That works well! thanks

DavidBruant commented:
The only two remaining things are the removal of `getSalesPayment` and removal of async/await

I'm going to try to work on `getSalesPayment` (and try to understand better how reallocate works first)

DavidBruant commented:
I removed async/awaits and i've rewritten so that the Opera uses payouts. For this i had the Opera creating the offers and redeeming invites directly (instead of having the contract do it)
As a consequence the contract code is smaller

I think this is ready for a review

katelynsills commented on packages/zoe/src/contracts/operaConcertTicket.js:
```diff
@@ -0,0 +1,138 @@
+
+  const inviteHandleByTicketNumber = new Map();
+
+  function completeAmountKeywordRecord(amountKeywordRecord) {
```
I think we can take this out, since it is recreating the amountKeywordRecord that exists as `offerRecord.amounts`.

katelynsills commented on packages/zoe/src/contracts/operaConcertTicket.js:
```diff
@@ -0,0 +1,138 @@
+    return harden(completed);
+  }
+
+  const auditoriumSeat = harden({
```
I think we can make this contract simpler by separating the creation of the ticket amounts from the actual minting of tickets. We can first create amounts for all of the tickets to be sold, and this can be what is returned from `getAvailableTickets`. That way, `getAvailableTickets` has nothing to do with offers or invites. Then, when a user chooses to purchase a ticket or a number of tickets, we can mint the tickets on demand, remove their extent from the `availableTickets` amount, create a new offer to escrow the tickets with Zoe, and then reallocate such that 1) the user's offer is allocated the tickets, 2) an offer representing the administrator is allocated the money, and 3) the new offer now is allocated nothing. This would follow the usage pattern we had in Autoswap. 

I can see why you had an administrative method to withdraw the money from the contract. To sum up the problem, it's that the seller of the tickets wants to receive a payout of money on demand, and doesn't want to wait until the tickets are sold to be able to receive the money? What if we simplify that constraint, and say that the seller of tickets won't withdraw their money until they are done selling tickets, regardless if they have all been sold or not. If we have an offer representing the entire ticket seller's position in the contract (not the short term offers created in the contract), then if that offer has the `exit` rule `onDemand`, when the ticket seller calls `cancel`, it would give the ticket seller their money and, if we write the contract to do so, shut down the contract.

DavidBruant replied:
> That way, getAvailableTickets has nothing to do with offers or invites. 

(`getAvailableTickets` only manipulates offerHandles, not invites... but i realize the code is confusing because of the inviteHandle/offerHandle mix. I'll fix the wording to only use `offerHandle`)

That's what i had initially written
and then i realized that keeping track of where the non-fungible tickets are and that none has been lost because of a bug and which assets belongs to whom (in the sense of "associated with which offerHandle") is the job Zoe+ERTP have been designed for
So i feel it's very relevant to leverage that instead of re-implementing it myself

And in order for Zoe+ERTP to keep track of what's still available, they need to be made aware of all tickets initially, hence minting ahead-of-time and not on-demand

I think this might be a place where the fact that the contract deals with non-fungible assets is a relevant difference with the Autoswap pattern (which certainly mints on-demand)

Finally, when all tickets are in the hands of Zoe and i have the corresponding offerHandles, i know i can trust Zoe+ERTP to tell me what is still available and what is not


> What if we simplify that constraint, and say that the seller of tickets won't withdraw their money until they are done selling tickets, regardless if they have all been sold or not.

Yes, exactly
This is what i was saying in an earlier comment

> 2) an offer representing the administrator is allocated the money

This part is hard with the contract minting the tickets
How much does the administrator trust the smart contract to perform an offer on its behalf (what you're proposing) rather than doing the offer themself after the contract produced invites (current version)?

I feel that rather than doing a 3-way offer which does work but is complicated and may lead to errors, the Opera is better to be left with the payments and invites, `redeem` early to make Zoe aware of the tickets and deal with the `payout` whenever they want (as the payment arrives or at the end; this option actually settles the question of the payouts because the Opera and not the contract is now responsible for collecting the payouts)

DavidBruant commented on packages/zoe/src/contracts/operaConcertTicket.js:
```diff
@@ -0,0 +1,138 @@
+      },
+      // This function returns a Map<TicketNumber, TicketExtent>
+      getAvailableTickets() {
+        return new Map(
```
because of current limitation of `@agoric/marshal`, a Map is inefficient
Change this to an array of extents

DavidBruant commented:
I rewrote the contract according to the discussion i've had with @katelynsills 
I think it's in a really good state

I'm particularly satisfied with the description of wanted allocations in `performExchange` where i used `amountMath.add`/`substract`.This is a lot clearer than what i was doing previously (making the amounts manually). This is possible because the auditorium position is represented by a single `offerHandle` instead of several ones as it was previously

katelynsills commented on packages/zoe/src/contracts/operaConcertTicket.js:
```diff
@@ -0,0 +1,174 @@
+            getAvailableTickets() {
+              // Because of a technical limitation in @agoric/marshal, an array of extents
+              // is better than a Map https://github.com/Agoric/agoric-sdk/issues/838
+              return zoe.getOffer(auditoriumOfferHandle).amounts.Ticket.extent;
```
This looks great!

DavidBruant commented on packages/zoe/test/unitTests/contracts/test-operaConcertTicket.js:
```diff
@@ -0,0 +1,409 @@
+                // The auditorium redeems its invite.
+                return (
+                  zoe
+                    .redeem(auditoriumInvite, harden({}))
```
oh... I'm realizing now that since the auditorium `want`s nothing, the buyers can get the tickets for free.
This is what we discussed yesterday and i wonder now whether that's a good idea. It sorts of offers zero guarantees for the auditorium since it's unable to express what hey want

Eventually, they'll be able to express the proportionality thing, but right now, it'd be good to have some level of guarantee as well

katelynsills replied:
Ah, sorry, I missed this comment and replied in the google doc. We can add a check to `performExchange` such that if the buyer is offering less than the ticket price, we reject the offer immediately. Further enforcement of offer safety, I think we can add in another PR or another tutorial.

DavidBruant replied:
> We can add a check to performExchange such that if the buyer is offering less than the ticket price, we reject the offer immediately.

I can do that, but I feel like this would be reimplementing what zoe is supposed to do for us

and that's what gives me the feeling that this contract is not a good example of all the safety Zoe could offer: as it is now, Zoe guarantees offer safety for the buyer, but not the auditorium. With your suggestion, Zoe+contract guarantees offer safety... but Zoe is supposed to be protecting against contract code malfunctions. I feel like the auditorium should be getting offer safety from Zoe alone, like the buyer gets

I'll wait for @erights review and procede with your suggestion if there's an agreement

erights replied:
Thanks for waiting and sorry my feedback took so long. As mentioned above, we do have a plan for Zoe enforcement of offer safety for this case. The so-called "multiples" offer I explain above. That's not going to happen soon. Until then, the auditorium's meaningful offer safety depends on the contract, not on Zoe.

Even after we add multiples, exclusive choice, offsets, and perhaps other things we find we need, there will always be an expressiveness limit to the offer safety Zoe itself can guarantee. The case that will probably be most painful soonest is offers in terms of "current" market prices, such as a "market order" entered into an exchange. We have no plans to have Zoe enforce safety relative to dynamic market prices at the time. This would always have to be delegated up to the contract.

Cases like market prices, which will come up a lot across many contracts, might cause another partitioning of risk within what Zoe regards as the contract level. @katelynsills did some good investigation of this early on, of generically parameterizable contract templates, that specialized code could be plugged into to create custom contracts. The templates could be highly vetted highly reusable frameworks. Such a template might be the code enforcing such higher-level safety properties, independent of what the specialized code is doing. This applies the Zoe risk partitioning principle at a higher layer of abstraction. The key thing is that multiple such templates can co-exist, each of which understand a different "language" for expressing such higher level offer safety constraints. No one such language need be universal.

I look forward to the day we get back to exploring that.

katelynsills commented on packages/zoe/test/unitTests/contracts/test-operaConcertTicket.js:
```diff
@@ -0,0 +1,409 @@
+          ticket => ticket.number === 1,
+        );
+        // make the corresponding amount
+        const ticket1Amount = ticketAmountMath.make(harden([ticket1Extent]));
```
Should these be amounts to start? I'm not sure of the answer

DavidBruant replied:
Either could work
I don't see a particular benefit to using one over the other in this context

katelynsills commented on packages/zoe/test/unitTests/contracts/test-operaConcertTicket.js:
```diff
@@ -0,0 +1,409 @@
+            try {
+              performExchange();
+            } catch (err) {
+              console.error('performaExchange', err);
```
Is this a typo?

DavidBruant replied:
yes

katelynsills commented on packages/zoe/test/unitTests/contracts/test-operaConcertTicket.js:
```diff
@@ -0,0 +1,409 @@
+                      t.equal(
+                        aliceBoughtTicketAmount.extent[0].number,
+                        1,
+                        'Alice should have receieved the ticket for the correct number',
```
```suggestion
                        'Alice should have received the ticket for the correct number',
```

katelynsills commented on packages/zoe/test/unitTests/contracts/test-operaConcertTicket.js:
```diff
@@ -0,0 +1,409 @@
+
+      const operaPurse = moolaIssuer.makeEmptyPurse();
+
+      const done = _operaPayout.then(payout => {
```
We've been reserving underscores for variables that won't be used. Since we are using the operaPayout, can we rename?

katelynsills commented on packages/zoe/src/contracts/operaConcertTicket.js:
```diff
@@ -0,0 +1,174 @@
+        const auditoriumSeat = harden({
+          // this is meant to be called right after redeem
+          // eventually, it might be done automatically: https://github.com/Agoric/agoric-sdk/issues/717
+          afterRedeem() {
```
I suspect we need this method, but can you walk me through why? It's because we cannot reallocate until the invite for the auditorium has been redeemed, right?

DavidBruant replied:
No, i think it's not needed actually. The code of this method can be moved to right after 
```js
const {
  invite: auditoriumInvite,
  inviteHandle: auditoriumOfferHandle,
} = zoe.makeInvite(auditoriumSeat);
```

because it doesn't need anything more than `auditoriumOfferHandle`

DavidBruant replied:
I'll move the code and remove the method

DavidBruant replied:
oh... I'm discovering that the method is needed
and for a reason that resonates with the discussion about [how people get `offerHandle`s](https://github.com/Agoric/agoric-sdk/issues/810#issuecomment-608460808)

Currently in `afterRedeem`, i get the current allocation for the `auditoriumOfferHandle`. And that's not possible to do that before the auditorium has redeemed the invite because calls to `zoe.getOffer(handle)` only works after `redeem`. It's only when redeem happens that the offerHandle is used for something useful:

https://github.com/Agoric/agoric-sdk/blob/e2c311b4d536f18d6966b0fc10721b581fd1368a/packages/zoe/src/zoe.js#L448-L458

(line 456 which uses the handle as key of a table entry that is consulted in `zoe.getOffer`)

I think this is further argument to create separate inviteHandle and offerHandle, and use each for its own purpose (inviteHandle for redeem+seat and offerHandle for interaction with allocations)

DavidBruant replied:
I'm going to keep the afterRedeem method for now

katelynsills replied:
Ok, sounds good, and thanks for the investigation.

DavidBruant commented:
> but I think after those are fixed, we are good to go

@katelynsills Have you seen https://github.com/Agoric/agoric-sdk/pull/780#discussion_r403360602 ?
I feel like the fact that anyone can buy tickets for free makes this contract a poor example of the safety guarantees that Zoe can provide if used correctly (unlike this contract with a `want` nothing proposal from the auditorium)

Before merging, i'd like to have your go on this point specifically because i've become skeptical of merging this

katelynsills commented:
> > but I think after those are fixed, we are good to go
> 
> @katelynsills Have you seen [#780 (comment)](https://github.com/Agoric/agoric-sdk/pull/780#discussion_r403360602) ?

Sorry, I missed the comment. I responded above. I think also, @erights should take a look at this and approve it before we merge

DavidBruant commented:
I addressed comments and change a bit how the tests are structured
I added Joker who is the person who's going to be doing things he shouldn't with the contract (and he should fail every time)
I added one test where Joker tries to buy a ticket for 1 moola (instead of 22). The tests expects him to fail, but as the code is now, he succeeds, so the test fails

There are 2 ways to fix the contract for the test to succeed (Joker fails to buy a ticket for 1 moola):
- Keep the auditorium position represented by 1 `want: nothing` offer and add a check in the contract (in `performExchange`) to reject the offer if the buyer payment is below 22 moolas
- Change the auditorium position to be represented by as many offers as there are tickets. Each offer is `want: 22 moolas` (Zoe does the offer safety check so it's not necessary to add something for this in the contract)

erights commented on packages/zoe/src/contracts/operaConcertTicket.js:
```diff
@@ -0,0 +1,174 @@
+    issuerKeywordRecord: { Money: moneyIssuer },
+  } = zoe.getInstanceRecord();
+
+  const moneyAmountMath = moneyIssuer.getAmountMath();
```
This doesn't look fixed yet. Unresolving. Or am I misunderstanding the review process flow we're using? Or did you just forget to push?

DavidBruant replied:
it isn't fixed, my mistake
I thought i had accepted the suggestion, but i hadn't

erights commented on packages/zoe/src/contracts/operaConcertTicket.js:
```diff
@@ -0,0 +1,174 @@
+    );
+    const ticketsPayment = mint.mintPayment(ticketsAmount);
+
+    const {
+      invite: contractSelfInvite,
+      inviteHandle: contractOfferHandle,
+    } = zoe.makeInvite();
+    // the contract creates an offer {give: tickets, want: nothing} with the tickets
+    return zoe
+      .getZoeService()
+      .redeem(
+        contractSelfInvite,
```
I see what you mean about invite making followed by immediate redemption being annoying unnecessary ceremony. Yes, we should fix this.

erights commented:
Do the first bulleted choice

>   * Keep the auditorium position represented by 1 want: nothing offer and add a check in the contract (in performExchange) to reject the offer if the buyer payment is below 22 moolas

not the second one. The second one better approximates the semantics of multiples, but I prefer not to do it that way until we have genuine multiples.

DavidBruant commented:
I had to rebase to take recent Zoe API changes into account
I added offer safety for the Auditorium in the contract

I think this is ready for a final review

DavidBruant commented:
I think we're good now

Proposition for squash-merge commit message

> feat(zoe): add Opera ticket contract, a non-fungible token contract

DavidBruant commented:
rebase on top of current master (with zoe helper import changes) + force-push
